### PR TITLE
fix(Core/DungeonFinder):

### LINKED ISSUE #5914: [Dungeon Finder] Requeueing after a run for the same dungeon should reset the dungeon
Original issue : https://github.com/chromiecraft/chromiecraft/issues/569

#### WHAT CLIENT DO YOU PLAY ON?

- [ ] enGB
- [x] enUS
- [ ] other (specify)

##### FACTION

- [x] Alliance
- [ ] Horde

##### CONTENT PHASE:

- [x] Generic
- [ ] 1-19
- [ ] 20-29
- [ ] 30-39
- [ ] 40-49
- [ ] 50-59

##### CURRENT BEHAVIOUR:

Queueing with the dungeon finder while being in a cleared dungeon doesn't reset the dungeon (thus the boss required for the completion of the run is missing).

##### EXPECTED BLIZZLIKE BEHAVIOUR:

It should teleport the group to the entrance (inside) and reset the whole dungeon.

##### STEPS TO REPRODUCE THE PROBLEM:

1. Queue for a dungeon finder.
2. Clear the dungeon.
3. Queue for the same dungeon as the one you've cleared (without teleporting out of the dungeon).
4. Accept. Notice that you didn't move to the entrance (inside) of the dungeon and mobs are still dead.

##### EXTRA NOTES:

Should happen with both factions.
Happened when we tried to queue for a random dungeon and our levels made that only one specific dungeon (Scarlet Monastery - Library) would happen.

##### AC HASH/COMMIT:

https://github.com/chromiecraft/azerothcore-wotlk/commit/6947789622de45479d2311fd5d1635af3a957b5d

##### OPERATING SYSTEM:

Ubuntu 20.04

##### MODULES:

- [mod-cfbg](https://github.com/azerothcore/mod-cfbg)
- [mod-duel-reset](https://github.com/azerothcore/mod-duel-reset)
- [mod-desertion-warnings](https://github.com/azerothcore/mod-desertion-warnings)
- [mod-ah-bot](https://github.com/azerothcore/mod-ah-bot)
- [mod-eluna-lua-engine](https://github.com/azerothcore/mod-eluna-lua-engine)
- [mod-server-auto-shutdown](https://github.com/azerothcore/mod-server-auto-shutdown)
- [lua-LevelUpReward](https://github.com/55Honey/Acore_LevelUpReward)
- [lua-CarbonCopy](https://github.com/55Honey/Acore_CarbonCopy)
- [mod-pvp-titles](https://github.com/azerothcore/mod-pvp-titles)

##### OTHER CUSTOMIZATIONS:

None.

##### SERVER:

ChromieCraft

<bountysource-plugin>

---
Want to back this issue? **[Post a bounty on it!](https://www.bountysource.com/issues/98542585-dungeon-finder-requeueing-after-a-run-for-the-same-dungeon-should-reset-the-dungeon?utm_campaign=plugin&utm_content=tracker%2F40032087&utm_medium=issues&utm_source=github)** We accept bounties via [Bountysource](https://www.bountysource.com/?utm_campaign=plugin&utm_content=tracker%2F40032087&utm_medium=issues&utm_source=github).
</bountysource-plugin>

### LINKED ISSUE #7388: Using RDF inside an instance, teleports you randomly
### What client do you play on?

enGB

### Faction

- [ ] Alliance
- [X] Horde

### Content Phase:

- [X] Generic
- [ ] 1-19
- [ ] 20-29
- [ ] 30-39
- [ ] 40-49
- [ ] 50-59

### Current Behaviour
Original report: https://github.com/chromiecraft/chromiecraft/issues/1360

If you are inside an instance, lets take WC; Fishing. Then your queue pops for an instance (Would imagine this also goes for BGS (Not Confirmed). After completing said dungeon you will be teleported out to a seemingly random spot in the world.

Hard to explain please see link below.
https://www.youtube.com/watch?v=nNqNzq8LluU

### Expected Blizzlike Behaviour

Not really sure, I would assume it teleported you to the entrance of the previous instance, in this case WC.

### Source

_No response_

### Steps to reproduce the problem

1. Enter an instance.
2. Go into RDF from said instance (Edited from original report)
3. Complete dungeon
4. Magic.

### Extra Notes
Testers notes:

I don't know whether it should teleport to prior instance, but random teleport seems wrong. At least, you should be teleported to the entrance of the prior instance.

### AC rev. hash/commit

https://github.com/chromiecraft/azerothcore-wotlk/commit/36a1777b99c400484537c358af8fa6b3a42b13c7

### Operating system

Ubuntu 20.04

### Modules

- [mod-ah-bot](https://github.com/azerothcore/mod-ah-bot)
- [mod-cfbg](https://github.com/azerothcore/mod-cfbg)
- [mod-chromie-xp](https://github.com/azerothcore/mod-chromie-xp)
- [mod-desertion-warnings](https://github.com/azerothcore/mod-desertion-warnings)
- [mod-duel-reset](https://github.com/azerothcore/mod-duel-reset)
- [mod-eluna-lua-engine](https://github.com/azerothcore/mod-eluna-lua-engine)
- [mod-ip-tracker](https://github.com/azerothcore/mod-ip-tracker)
- [mod-low-level-arena](https://github.com/azerothcore/mod-low-level-arena)
- [mod-multi-client-check](https://github.com/azerothcore/mod-multi-client-check)
- [mod-pvp-titles](https://github.com/azerothcore/mod-pvp-titles)
- [mod-pvpstats-announcer](https://github.com/azerothcore/mod-pvpstats-announcer)
- [mod-queue-list-cache](https://github.com/azerothcore/mod-queue-list-cache)
- [mod-server-auto-shutdown](https://github.com/azerothcore/mod-server-auto-shutdown)
- [lua-CarbonCopy](https://github.com/55Honey/Acore_CarbonCopy)
- [lua-CustomWorldboss](https://github.com/55Honey/Acore_CustomWorldboss)
- [lua-LevelUpReward](https://github.com/55Honey/Acore_LevelUpReward)
- [lua-send-and-bind](https://github.com/55Honey/Acore_SendAndBind)
- [lua-Zonecheck](https://github.com/55Honey/acore_Zonecheck)
- [lua-TempAnnouncements](https://github.com/55Honey/Acore_TempAnnouncements)

### Customizations

None

### Server

ChromieCraft

### PR DESCRIPTION
**Pull request by UltraNix**
Players should be able to enter the same dungeon if just completed it.
Properly define if proposal is new.
Set teleport point to nearest graveyard if teleported to dungeon from other instance.
Display completed encounters when joining LFG in progress. (excluding random).
Fixed #5914
Fixed #7388

## Issues Addressed:

- Closes #5914
- Closes #7388
- Closes https://github.com/chromiecraft/chromiecraft/issues/569
- Closes https://github.com/chromiecraft/chromiecraft/issues/1360

## Tests Performed:

- Tested ingame

## How to Test the Changes:

More info [here](#5914) and [here](#7388)

## How to Test AzerothCore PRs
 
When a PR is ready to be tested, it will be marked as **[WAITING TO BE TESTED]**.

You can help by testing PRs and writing your feedback here on the PR's page on GitHub. Follow the instructions here:

http://www.azerothcore.org/wiki/How-to-test-a-PR

**REMEMBER**: when testing a PR that changes something **generic** (i.e. a part of code that handles more than one specific thing), the tester should not only check that the PR does its job (e.g. fixing spell XXX) but **especially** check that the PR does not cause any regression (i.e. introducing new bugs).

**For example**: if a PR fixes spell X by changing a part of code that handles spells X, Y, and Z, we should not only test X, but **we should test Y and Z as well**.

### CODE REVIEW

**Comment by pklloveyou** [comment]
If all dungeons are locked, you can't enter...

**Comment by UltraNix** [comment]
So it's blizzlike.

**Comment by pklloveyou** [comment]
Is it correct not to reset the instance?

**Comment by UltraNix** [comment]
If it's a new instance - it should reset.

**Comment by pklloveyou** [comment]
If so, then tested

**Comment by pklloveyou** [comment]
```
        // pussywizard:
        if (isRaid && grp && (grp->isLFGGroup() || guid != grp->GetLeaderGUID()))
            return;

+        //Do not allow to change dungeon in the middle of a current dungeon
+        if (isContinue)
+        {
+            dungeons.clear();
+            dungeons.insert(GetDungeon(gguid));
+            joinData.result = LFG_JOIN_PARTY_NOT_MEET_REQS;
+        }

        // Can't join. Send result
        if (joinData.result != LFG_JOIN_OK)
```

**Comment by pklloveyou** [comment]
After the player escapes from the dungeon, the new player who joins the dungeon again is not in the same instance as the dungeon player..

File: src/server/game/DungeonFinding/LFGMgr.cpp [hunk 1]
```diff
@@ -1603,6 +1603,26 @@ namespace lfg
 
         _SaveToDB(gguid);
 
+        // Select a player inside to be teleported to
+        WorldLocation const* teleportLocation = nullptr;
+        for (GroupReference* itr = grp->GetFirstMember(); itr != nullptr; itr = itr->next())
+        {
+            Player* plr = itr->GetSource();
+            if (plr && plr->GetMapId() == uint32(dungeon->map))
```
Comment (pklloveyou) [inline src/server/game/DungeonFinding/LFGMgr.cpp:1611]:
```

        // Select a player inside to be teleported to
        WorldLocation const* teleportLocation = nullptr;
        Player* leader = ObjectAccessor::FindPlayerInOrOutOfWorld(grp->GetLeaderGUID());
        for (GroupReference* itr = grp->GetFirstMember(); itr != nullptr; itr = itr->next())
        {
            Player* plr = itr->GetSource();
            if (plr && plr->GetMapId() == uint32(dungeon->map) && plr->GetInstanceId() != leader->GetInstanceId())

```

File: src/server/game/DungeonFinding/LFGMgr.cpp [hunk 2]
```diff
@@ -1603,6 +1603,26 @@ namespace lfg
...
+            {
+                // Remove bind to that map
+                if (!sInstanceSaveMgr->PlayerIsPermBoundToInstance(plr->GetGUID(), dungeon->map, plr->GetDungeonDifficulty()))
```
Comment (pklloveyou) [inline src/server/game/DungeonFinding/LFGMgr.cpp:1614]:
```c++

        WorldLocation const* teleportLocation = nullptr;
+        Player* leader = ObjectAccessor::FindPlayerInOrOutOfWorld(grp->GetLeaderGUID());
        for (GroupReference* itr = grp->GetFirstMember(); itr != nullptr; itr = itr->next())
        {
            Player* plr = itr->GetSource();
            if (plr && plr->GetMapId() == uint32(dungeon->map))
            {
+                bool UnbindInstance = leader && plr->GetInstanceId() != leader->GetInstanceId();
                // Remove bind to that map
-                if (!sInstanceSaveMgr->PlayerIsPermBoundToInstance(plr->GetGUID(), dungeon->map, plr->GetDungeonDifficulty()))
+                if (UnbindInstance && !sInstanceSaveMgr->PlayerIsPermBoundToInstance(plr->GetGUID(), dungeon->map, plr->GetDungeonDifficulty()))


```

Comment (pklloveyou) [reply src/server/game/DungeonFinding/LFGMgr.cpp:1614]: Fix.
After the player escapes from the dungeon, the new player who joins the dungeon again is not in the same instance as the dungeon player..

**Comment by pklloveyou** [comment]
GroupHandler.cpp
void WorldSession::HandleGroupInviteOpcode(WorldPacket& recvData)
```c++
    Group* group = GetPlayer()->GetGroup();
+    if (group && group->isLFGGroup() && group->IsLfgRandomInstance())
+    {
+        SendPartyResult(PARTY_OP_INVITE, membername, ERR_TARGET_NOT_IN_INSTANCE_S);
+        return;
+    }
    if (group && (group->isBGGroup() || group->isBFGroup()))
```
Random Dungeon team. The captain is not allowed to invite players

File: src/server/game/DungeonFinding/LFGMgr.cpp (hunk 2 above)
Comment (UltraNix) [reply src/server/game/DungeonFinding/LFGMgr.cpp:1614]: So with this fix the players' binds are never cleared, so always been on different instance id. `plr->GetInstanceId()` is always the same as `leader->GetInstanceId()` if being on the same.

**Comment by UltraNix** [comment]
> GroupHandler.cpp
> void WorldSession::HandleGroupInviteOpcode(WorldPacket& recvData)
> 
> ```c++
>     Group* group = GetPlayer()->GetGroup();
> +    if (group && group->isLFGGroup() && group->IsLfgRandomInstance())
> +    {
> +        SendPartyResult(PARTY_OP_INVITE, membername, ERR_TARGET_NOT_IN_INSTANCE_S);
> +        return;
> +    }
>     if (group && (group->isBGGroup() || group->isBFGroup()))
> ```
> 
> Random Dungeon team. The captain is not allowed to invite players
> Otherwise, the Random Dungeon team will be destroyed

Why you want to invite players? It's DF.

**Comment by pklloveyou** [comment]
> > GroupHandler.cpp
> > void WorldSession::HandleGroupInviteOpcode(WorldPacket& recvData)
> > ```c++
> >     Group* group = GetPlayer()->GetGroup();
> > +    if (group && group->isLFGGroup() && group->IsLfgRandomInstance())
> > +    {
> > +        SendPartyResult(PARTY_OP_INVITE, membername, ERR_TARGET_NOT_IN_INSTANCE_S);
> > +        return;
> > +    }
> >     if (group && (group->isBGGroup() || group->isBFGroup()))
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > Random Dungeon team. The captain is not allowed to invite players
> > Otherwise, the Random Dungeon team will be destroyed
> 
> Why you want to invite players? It's DF.

https://github.com/azerothcore/azerothcore-wotlk/pull/7488#issuecomment-903651816

File: src/server/game/DungeonFinding/LFGMgr.cpp (hunk 2 above)
Comment (pklloveyou) [reply src/server/game/DungeonFinding/LFGMgr.cpp:1614]: Well,  but it must be wrong not to be in the same instance

**Comment by UltraNix** [comment]
Okay - I get it.

File: src/server/game/DungeonFinding/LFGMgr.cpp (hunk 2 above)
Comment (UltraNix) [reply src/server/game/DungeonFinding/LFGMgr.cpp:1614]: I tested that and never encountered such a problem. As players has been teleported to the same instance. Can you tell me what to do to trigger that bug (step by step)?

Comment (pklloveyou) [reply src/server/game/DungeonFinding/LFGMgr.cpp:1614]: 1. One player leaves the random team
2. New players join the random queue, enter the random dungeon, and then can't see other players

Comment (pklloveyou) [reply src/server/game/DungeonFinding/LFGMgr.cpp:1614]: Another problem is that if the last "new player" binds this instance before joining the random dungeon, the dungeon Finder will not add the "new player" to the instance of this dungeon

Comment (UltraNix) [reply src/server/game/DungeonFinding/LFGMgr.cpp]: @pklloveyou Should be fixed.

**Comment by pklloveyou** [comment]
Tested

File: src/server/game/DungeonFinding/LFGQueue.cpp [hunk 3]
```diff
@@ -429,8 +432,27 @@ namespace lfg
...
             if (!proposal.isNew && data.group && data.group == proposal.group) // Player from existing group, autoaccept
                 data.accept = LFG_ANSWER_AGREE;
+
+            if (!completedEncounters && !proposal.isNew)
+            {
+                if (LFGDungeonEntry const* dungeon = sLFGDungeonStore.LookupEntry(proposal.dungeonId))
+                {
+                    if (Player* player = ObjectAccessor::FindConnectedPlayer(itRoles->first))
+                    {
+                        if (player->GetMapId() == dungeon->map)
```
Comment (Faq) [inline src/server/game/DungeonFinding/LFGQueue.cpp:442]: fatal error: comparison of integers of different signs: 'uint32' (aka 'unsigned int') and 'const int32' (aka 'const int') [-Wsign-compare]

**Comment by pklloveyou** [comment]
image
Is it a mistake to crash here?

**Comment by UltraNix** [comment]
No. As you see the log: player cannot enter the map if already there.

File: src/server/game/DungeonFinding/LFGQueue.cpp (hunk 3 above)
Comment (FrancescoBorzi) [reply src/server/game/DungeonFinding/LFGQueue.cpp]: @UltraNix @faq has this been solved?

**Comment by pklloveyou** [comment]
ok

File: src/server/game/DungeonFinding/LFGMgr.cpp [hunk 4]
```diff
@@ -1603,21 +1611,44 @@ namespace lfg
...
+
         bool randomDungeon = false;
         // Teleport Player
         for (LfgGuidList::const_iterator it = playersToTeleport.begin(); it != playersToTeleport.end(); ++it)
             if (Player* player = ObjectAccessor::FindPlayer(*it))
             {
                 if (player->GetGroup() != grp) // pussywizard: could not add because group was full (some shitness happened)
                     continue;
+
+                if (player->GetMapId() == uint32(dungeon->map))
```
Comment (Kitzunu) [inline src/server/game/DungeonFinding/LFGMgr.cpp:1634]: I see that further down you use static_cast and here not. What is the difference between the different cases?

Comment (UltraNix) [reply src/server/game/DungeonFinding/LFGMgr.cpp:1634]: None.  `uint32` cast is c style.

Comment (Kitzunu) [reply src/server/game/DungeonFinding/LFGMgr.cpp:1634]: yea, I just want to learn why here you use `player->GetMapId() == uint32(dungeon->map)` and in a line below is `player->GetMapId() == static_cast<uint32>(dungeon->map)`

File: src/server/game/DungeonFinding/LFGQueue.cpp (hunk 3 above)
Comment (UltraNix) [reply src/server/game/DungeonFinding/LFGQueue.cpp]: Yes

File: src/server/game/DungeonFinding/LFGMgr.cpp (hunk 4 above)
Comment (UltraNix) [reply src/server/game/DungeonFinding/LFGMgr.cpp:1634]: Just laziness :P
Both doing the same.

**Comment by FrancescoBorzi** [comment]
can this be merged?

**Comment by pklloveyou** [comment]
It's working on my server

**Comment by IntelligentQuantum** [comment]
It works fine for me

**Review by IntelligentQuantum — APPROVED** [review]

**Pull request merged**